controllers/etcdpeer_controller_test.go

```diff
@@ -0,0 +1,62 @@
+package controllers
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
```

Comment:
what was the reasoning for choosing these test libs? i've personally had some bad experience with them. imo ginko is usually no more powerful than properly written vanilla golang tests + gomega is usually better replaced with parts of the testify suite.  like a lot of the k8s codebase, they seem to be java idioms hacked in to golang's syntax, and the alternatives are usually much more accessible for a typical golang dev.

Reply:
> [W]hat was the reasoning for choosing these test libs?

They're created by the default Kubebuilder scaffolding. Kubebuilder seems to have had some discussion about this themselves, as [summarised by one of their developers](https://github.com/kubernetes-sigs/kubebuilder/issues/822#issuecomment-506887995). The short version is that they've found Ginkgo and Gomega easier to write cleaner tests in than default Go and it's harder to 'get wrong'. But, they've expressed that they want to switch to a customisable test generator.

My main concern with moving to 'vanilla' Go tests would be that it might make it harder to onboard people into the project. It means that we'd be diverging from Kubebuilder's expected layout, which might confuse people who are reading the Kubebuilder book or comparing to other Kubebuilder codebases. We'd also loose some scaffolding abilities (e.g., when adding new resources).

Personally I have a mild preference for this behavioural kind of test, although I've not used Ginkgo and Gomega before myself. @munnerz did you have any thoughts on this?

Reply:
> Go tests would be that it might make it harder to onboard people into the project

i'd argue that the set of golang devs who understand vanilla tests is much larger than the set of devs who are familiar with ginko/gomega. that's especially true within Improbable, who will be primarily responsible for maintaining this project

> they've found Ginkgo and Gomega easier to write cleaner tests

as i understand it, k8s was mostly developed by the google engineers who were working on Borg, a lot of them did not have golang experience beforehand so the project ended up being heavily influenced by java idioms, this seems like one of them.  i'll push up a branch containing a rewrite of these tests using vanilla feautes + stretchr/testify for assertions, and we can see what we prefer.

Reply:
:+1: I think keeping this as is for now makes sense, given we are trying to get to a point where we can get clusters up and running, and the majority of the examples and sample material utilises this.

That said I'm very very open to a vanilla go testing approach, and +1 to get going on rewriting things to use it.

> i'll push up a branch containing a rewrite of these tests using vanilla feautes + stretchr/testify for assertions, and we can see what we prefer.

This sounds like the best approach to me 😄 (get this one in, keep moving forward, and then work on the rewrite - we have very few test cases now, so shouldn't be too painful 😄)

Reply:
> that's especially true within Improbable, who will be primarily responsible for maintaining this project

It's a fair point.

> …a lot of them did not have golang experience beforehand so the project ended up being heavily influenced by java idioms, this seems like one of them. 

A lot of these behavioural test frameworks, tarred by the Java brush as they may be, are actually quite good (but I could be [biased](https://medium.com/greshamtech/testing-java-applications-with-groovy-spock-49b4fd68ddbc)). But that's more of a theoretical conversation.

My preference isn't all that strong. If we feel that vanilla Go testing would be better then I'm happy to move to that. I'd prefer to, as James says, merge this now and open another PR to migrate to that if possible. But if you'd like to hold this one until we make those changes it's not actually a huge issue for me (as my bootstrap branch is based on this one anyway).

Reply:
happy to merge this in and revisit once i get a chance to demonstrate the rewrite, if it will unblock you.